contracts/upgradeable_contracts/BasicTokenBridge.sol

```diff
@@ -7,95 +7,180 @@ import "./Ownable.sol";
 
     function withinExecutionLimit(uint256 _amount) public view returns (bool) {
-        uint256 nextLimit = totalExecutedPerDay(getCurrentDay()).add(_amount);
-        return executionDailyLimit() >= nextLimit && _amount <= executionMaxPerTx();
+        return _getWithinLimit(GET_WITHIN_EXECUTION_LIMIT, _amount, _getTokenBalance());
     }
 
-    function getCurrentDay() public view returns (uint256) {
-        // solhint-disable-next-line not-rely-on-time
-        return now / 1 days;
+    function getTotalSpentPerDay(uint256 _day) public view returns (uint256) {
+        return _getUint(GET_TOTAL_SPENT_PER_DAY, _day);
     }
 
-    function setTotalSpentPerDay(uint256 _day, uint256 _value) internal {
-        uintStorage[keccak256(abi.encodePacked("totalSpentPerDay", _day))] = _value;
+    function getTotalExecutedPerDay(uint256 _day) public view returns (uint256) {
+        return _getUint(GET_TOTAL_EXECUTED_PER_DAY, _day);
     }
 
-    function setTotalExecutedPerDay(uint256 _day, uint256 _value) internal {
-        uintStorage[keccak256(abi.encodePacked("totalExecutedPerDay", _day))] = _value;
+    function _increaseTotalSpentPerDay(uint256 _amount) internal {
+        _execute(INCREASE_TOTAL_SPENT_PER_DAY, _amount);
     }
 
-    function setDailyLimit(uint256 _dailyLimit) external onlyOwner {
-        require(_dailyLimit > maxPerTx() || _dailyLimit == 0);
-        uintStorage[DAILY_LIMIT] = _dailyLimit;
-        emit DailyLimitChanged(_dailyLimit);
+    function _increaseTotalExecutedPerDay(uint256 _amount) internal {
+        _execute(INCREASE_TOTAL_EXECUTED_PER_DAY, _amount);
     }
 
-    function setExecutionDailyLimit(uint256 _dailyLimit) external onlyOwner {
-        require(_dailyLimit > executionMaxPerTx() || _dailyLimit == 0);
-        uintStorage[EXECUTION_DAILY_LIMIT] = _dailyLimit;
-        emit ExecutionDailyLimitChanged(_dailyLimit);
+    function _updateTodayLimit() internal {
+        _execute(UPDATE_TODAY_LIMIT, _getTokenBalance());
     }
 
-    function setExecutionMaxPerTx(uint256 _maxPerTx) external onlyOwner {
-        require(_maxPerTx < executionDailyLimit());
-        uintStorage[EXECUTION_MAX_PER_TX] = _maxPerTx;
+    function _setLimits(uint256[] _requestLimitsArray, uint256[] _executionLimitsArray) internal {
+        require(limitsContract().delegatecall(abi.encodeWithSelector(SET_LIMITS, _requestLimitsArray, _executionLimitsArray)));
     }
 
-    function setMaxPerTx(uint256 _maxPerTx) external onlyOwner {
-        require(_maxPerTx < dailyLimit());
-        uintStorage[MAX_PER_TX] = _maxPerTx;
+    function _execute(bytes4 _method, uint256 _value) internal {
+        require(limitsContract().delegatecall(abi.encodeWithSelector(_method, _value)));
     }
 
-    function setMinPerTx(uint256 _minPerTx) external onlyOwner {
-        require(_minPerTx < dailyLimit() && _minPerTx < maxPerTx());
-        uintStorage[MIN_PER_TX] = _minPerTx;
+    function _getTokenBalance() internal view returns (uint256) {}
```

Comment:
Do you think we could declare `_getTokenBalance` as an abstract method instead of an empty one?
```suggestion
    function _getTokenBalance() internal view returns (uint256);
```